### DISCUSSION TITLE
Parameterized query shows NVARCHAR even for VARCHAR column?

### DISCUSSION THREAD

**Discussion by D00mSlayer** [sqlalchemy/sqlalchemy#10524] [Usage Questions] [answered]
### Observation
I have a table T_VARCHAR with columns - id (INT) , name (VARCHAR SQL_Latin1_General_CP1_CI_AS)

I observed that NVARCHAR is used when querying a VARCHAR column in SQL Server (got information from SQL Profiler).
In my example below, I am inserting simple ASCII character set to observe the behavior.

### Profiler generated query

1. Using PyODBC (default settings) inserting ASCII to T_VARCHAR
    > declare @p1 int
    > set @p1=4
    > exec sp_prepexec @p1 output,N'@P1 **nvarchar**(188)',N'INSERT INTO t_varchar (name) OUTPUT inserted.id VALUES (@P1)',N'!"#$%&''()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
    > select @p1

2. Using PyODBC (setencoding property to utf-8) inserting ASCII to T_VARCHAR
    > declare @p1 int
    > set @p1=4
    > exec sp_prepexec @p1 output,N'@P1 **varchar**(94)',N'INSERT INTO t_varchar (name) OUTPUT inserted.id VALUES (@P1)','!"#$%&''()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
    > select @p1

### Expectation
Since the column is VARCHAR, shouldn't SQL Server use VARCHAR and not NVARCHAR? Also, regarding the collate **SQL_Latin1_General_CP1_CI_AS**, I understand that SQL Server will try to find the closest match with the corresponding code page for any given character; which is [CP1252](https://en.wikipedia.org/wiki/Windows-1252). If it does not find any, it would just replace with `?`.

### Databases / Backends / Drivers targeted

- SQL Server 2022
- Python - 3.10
- pyodbc - 4.0.35
- schematics - 2.1.1
- SQLAlchemy - 1.3.17
- ODBC Driver - [Microsoft ODBC 17](https://learn.microsoft.com/en-us/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server?view=sql-server-ver16&tabs=alpine18-install%2Calpine17-install%2Cdebian8-install%2Credhat7-13-install%2Crhel7-offline#17)

### Example Use

I have attached a github link with the code. Please have a look [HERE](https://github.com/D00mSlayer/issue-nvarchar-cast).

### Additional context

A very similar issue was raised years ago: https://github.com/sqlalchemy/sqlalchemy/issues/5345

**Comment by CaselIT** [comment]
Hi,

I'm not sure what you are doing, since no example is provided.

> In my example below, I am inserting simple ASCII character set to observe the behavior.

That does not seem anything generated by sqlalchemy. If you are using strings you need to indicate the types of the parameters.

Using a sqlalchemy query it correctly uses varchars are parameter types

```py
import sqlalchemy as sa
from sqlalchemy import event

meta = sa.MetaData()
tbl = sa.Table(
    "t_vchar",
    meta,
    sa.Column("id", sa.Integer, primary_key=True),
    sa.Column("name", sa.String(200, "SQL_Latin1_General_CP1_CI_AS")),
)

e = sa.create_engine(
    "mssql+pyodbc://scott:tiger^5HHH@127.0.0.1:1433/test?driver=ODBC+Driver+18+for+SQL+Server&TrustServerCertificate=yes",
    echo=True,
)
with e.connect() as c:
    meta.drop_all(c)
    meta.create_all(c)

    @event.listens_for(e, "do_setinputsizes")
    def go(inputsizes, cursor, statement, parameters, context):
        print(statement, inputsizes, parameters)

    c.execute(tbl.insert(), {"name": "foo"})
```

**Reply by D00mSlayer** [reply]
I have posted a github link with my example. However, I see the same issue with your sample as well. I just generated the logs from SQL Server:

> declare @p1 int
> set @p1=3
> exec sp_prepexec @p1 output,N'@P1 nvarchar(6)',N'INSERT INTO t_vchar (name) OUTPUT inserted.id VALUES (@P1)',N'foo'
> select @p1

**Reply by gordthompson** [reply]
> SQLAlchemy - 1.3.17

setinputsizes support was added in SQLAlchemy 1.4

**Reply by D00mSlayer** [reply]
@gordthompson I can see you have a lot of knowledge about this particular issue. For now, I am listening to before_execute event to setencoding (utf-8) and setdecoding on pyodbc.SQL_CHAR AND pyodbc.SQL_WCHAR. With this I see indexes being utilised.

**Reply by gordthompson** [reply]
With SQLAlchemy 2.0.22, this code

```python
from sqlalchemy import Column, Integer, String, create_engine, select
from sqlalchemy.orm import Session, declarative_base

engine = create_engine("mssql+pyodbc://scott:tiger^5HHH@mssql_199")
Base = declarative_base()


class Thing(Base):
    __tablename__ = "thing"
    id = Column(Integer, primary_key=True)
    txt = Column(String(50, "SQL_Latin1_General_CP1_CI_AS"))


Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

with Session(engine) as sess:
    sess.add(Thing(txt="testing"))
    sess.commit()
    my_thing = sess.execute(select(Thing).where(Thing.txt == "testing")).all()
```

produces the following output in SQL Profiler:

```
declare @p1 int
set @p1=3
exec sp_prepexec @p1 output,N'@P1 varchar(14)',N'INSERT INTO thing (txt) OUTPUT inserted.id VALUES (@P1)','testing'
select @p1

declare @p1 int
set @p1=4
exec sp_prepexec @p1 output,N'@P1 varchar(14)',N'SELECT thing.id, thing.txt 
FROM thing 
WHERE thing.txt = @P1','testing'
select @p1
```

Note that in both cases the parameter is declared as `varchar` (`@P1 varchar(14)`) and the value is sent as a `varchar` value (`'testing'`), not as an `nvarchar` value (`N'testing'`).

**Reply by D00mSlayer** [reply]
I see and understand that. We are currently restricted to v1.3.17 and don't have much wiggle room.

**Reply by CaselIT** [reply]
Just curious, why .17 and not the last one of the 1.3 series (.24)?

Late number usually just include bug fixes

**Reply by D00mSlayer** [reply]
@CaselIT We could probably do that. Need to do impact analysis But with .24, is there any change that could fix my issue?

**Reply by CaselIT** [reply]
probably not, you would have to look at the changelog.
As mentioned I was just curious why that oddly specific version of 1.3

**Reply by D00mSlayer** [reply]
@CaselIT : checked the changelog and I did not find any reference with regards to my issue. It seems that using setencoding and setdecoding to utf-8 will have issues with the existing data. Probably requires a data migration.

**Reply by D00mSlayer** [reply]
What would you suggest I do? @gordthompson @CaselIT 
Should I subscribe to connection event and setencoding / setdecoding to utf-8 or utilize setinputsizes? Again, upgrading SQLAlchemy isn't possible at the moment for our application.

**Reply by gordthompson** [reply, marked as answer]
Using setencoding/setdecoding for SQL Server is not recommended. If you use

```python
cnxn.setencoding('utf-8')
cnxn.setdecoding(pyodbc.SQL_CHAR, 'utf-8')
```

then your varchar strings will round-trip, but only because they get mangled on the way in and un-mangled on the way out. In the database itself, `sess.add(Thing(txt="fermé"))` will result in the string being stored as `fermÃ©`.

If you are pinned to SQLA 1.3, then I'd suggest that you do INSERTs and SELECTs the normal way and let SQL Server handle the mapping between Unicode and SQL_Latin1_General_CP1_CI_AS. If you have performance issues with Unicode parameters in a WHERE clause (causing index scans instead of index seeks) then you can do

```python
    my_thing = sess.scalar(
        select(Thing).where(Thing.txt == func.cast("fermé", String(50)))
    )
```

which results in SQLAlchemy emitting

```
SELECT thing.id, thing.txt 
FROM thing 
WHERE thing.txt = CAST(? AS VARCHAR(50))
[generated in 0.00045s] ('fermé',)
```

and SQL Profiler showing

```
declare @p1 int
set @p1=4
exec sp_prepexec @p1 output,N'@P1 nvarchar(10)',N'SELECT thing.id, thing.txt 
FROM thing 
WHERE thing.txt = CAST(@P1 AS VARCHAR(50))',N'fermé'
select @p1```

**Reply by D00mSlayer** [reply]
Thank you @gordthompson 
Although changing code throughout the app is going to be a pain. I would obviously want to upgrade SQLAlchemy but there are too many changes between 1.3.17 and 2.0.x

**Reply by gordthompson** [reply]
Understood. Note that my examples above were wrong. I have fixed them.
